### PR TITLE
NestedFolderPicker: Implement keyboard handling

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- implements keyboard selection behaviour in the nested folder picker
- implements correct aria attributes to properly express the tree

**Why do we need this feature?**

- so the new `FolderPicker` is accessible to all users

**Who is this feature for?**

- everyone!

**Which issue(s) does this PR fix?**:

For https://github.com/grafana/grafana/issues/65745

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/core/components/NestedFolderPicker/NestedFolderList.tsx [hunk 1]
```diff
@@ -68,39 +80,29 @@ interface RowProps {
...
 
 function Row({ index, style: virtualStyles, data }: RowProps) {
-  const { items, foldersAreOpenable, selectedFolder, onFolderClick, onSelectionChange } = data;
-  const { item, isOpen, level } = items[index];
+  const { items, focusedItemIndex, foldersAreOpenable, selectedFolder, onFolderExpand, onFolderSelect } = data;
+  const { item, isOpen, level, parentUID } = items[index];
+  const rowRef = useRef<HTMLDivElement>(null);
+  const labelId = useId();
+  const children = items.filter((i) => i.parentUID === item.uid).map((i) => i.item.uid);
+  const posInSet = items.filter((i) => i.parentUID === parentUID).findIndex((i) => i.item.uid === item.uid) + 1;
```
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:88]: don't think we'll need to do this once we share state from redux -> we can use the `childrenByParentUID` selector

File: public/app/core/components/NestedFolderPicker/NestedFolderList.tsx [hunk 2]
```diff
@@ -111,25 +113,36 @@ function Row({ index, style: virtualStyles, data }: RowProps) {
...
+        [styles.rowSelected]: item.uid === selectedFolder,
+      })}
+      tabIndex={-1}
+      onClick={handleSelect}
+      aria-expanded={isOpen}
+      aria-selected={item.uid === selectedFolder}
+      aria-labelledby={labelId}
+      aria-level={level + 1} // aria-level is 1-indexed
+      role="treeitem"
+      aria-owns={children.length > 0 ? children.map((child) => getDOMId(child)).join(' ') : undefined}
```
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:132]: this is ugly but i don't think there's another way 😬 i was hoping there was a way to specify a parent from the child, but looking through the spec it looks like each parent must explicitly list all it's children 🤮

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 3]
```diff
@@ -123,39 +117,105 @@ export function NestedFolderPicker({ value, onChange }: NestedFolderPickerProps)
...
+    },
+    [onChange]
+  );
+
   const isLoading = rootFoldersState.loading || searchState.loading;
   const error = rootFoldersState.error || searchState.error;
 
   const tree = flatTree;
 
+  const handleKeyDown = useCallback(
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:164]: We should put a bunch of these hooks in like a `useTreeInteractions` hooks in a seperate file to help readability.

**Review by joshhunt — COMMENTED** [review]
few quick thoughts as i test this

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 4]
```diff
@@ -123,39 +117,105 @@ export function NestedFolderPicker({ value, onChange }: NestedFolderPickerProps)
...
+    (ev: React.KeyboardEvent<HTMLInputElement>) => {
+      const foldersAreOpenable = !(search && searchState.value);
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:166]: lets hoist this up and reuse it here and down where we pass it into NestedFolderList

File: public/app/core/components/NestedFolderPicker/NestedFolderList.tsx [hunk 5]
```diff
@@ -15,31 +15,43 @@ import { FolderUID } from './types';
 
 const ROW_HEIGHT = 40;
 const CHEVRON_SIZE = 'md';
+const ID_PREFIX = 'folder-picker-item';
+
+export const getDOMId = (id: string) => `${ID_PREFIX}-${id ?? 'root'}`;
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:20]: the one 'problem' (that isn't actually a problem) is that these IDs are shared across all instances of NestedFolderPicker (they aren't globally unique), but you can't have multiple folder pickers open at the same time so in practice I think this isn't a real issue.

**Review by joshhunt — APPROVED** [review]
Really great stuff on this. I was pleasantly surprised to see that left and right arrows still let to move your cursor when searching :)

**Comment by ashharrison90** [comment]
@joshhunt have ensured the ids are unique and refactored out the keyboard logic into a custom hook... i'll wait for your PR to go in first, merge with that and ask for a rereview then 👍

**Review by joshhunt — APPROVED** [review]
again, well done on this! 👏🏻

**Pull request merged**